README.md
# Application Localizer

Application Localizer package that helps with localizing applications

[Pseudo Locale Generator Web Site](https://ppatotski.github.io/app-localizer/)

[![NPM version](https://img.shields.io/npm/v/app-localizer.svg)](https://www.npmjs.com/package/app-localizer)
[![Build Status](https://travis-ci.org/ppatotski/app-localizer.svg?branch=master)](https://travis-ci.org/ppatotski/app-localizer)
[![Coverage Status](https://coveralls.io/repos/github/ppatotski/app-localizer/badge.svg?branch=master)](https://coveralls.io/github/ppatotski/app-localizer?branch=master)

> uses [Intl MessageFormat Parser](https://github.com/yahoo/intl-messageformat-parser) that parses ICU Message strings into an AST.
>
> tested with [Intl MessageFormat](https://github.com/yahoo/intl-messageformat) that formats ICU Message strings with number, date, plural, and select placeholders to create localized messages.
>
> used by [vscode-app-localizer](https://github.com/ppatotski/vscode-app-localizer) vscode extension
>
> [![VSCode Extension Release](http://vsmarketplacebadge.apphb.com/version/gsppvo.vscode-app-localizer.svg)](https://marketplace.visualstudio.com/items?itemName=gsppvo.vscode-app-localizer)

## Features

* Locale validator

  * Check for missing labels
  * Multi-file locale support (see [example](#multi-file-locale-example) below)
  * Polymer file structure
  * Angular flat file structure

* Pseudo locale generator (char mapping is taken from [pseudolocalization-tool](https://code.google.com/archive/p/pseudolocalization-tool/))
  * Accents on letters
  * Longer sentence
  * Longer word
  * Right-to-Left
  * Enclose in exclamations
  * Enclose in brackets - Support [ICU Message syntax](https://formatjs.io/guides/message-syntax/)
* Cross-platform

## Install

```shell
npm install app-localizer
```

## API Reference

Pseudo locale generator options

* **expander** Sentence expand factor 0.3 = 30%
* **wordexpander** Word expand factor 0.5 = 50%
* **brackets** Enclose sentence in brackets
* **exclamations** Enclose sentence in exclamations
* **accents** Convert letter to its accent version
* **rightToLeft** RTL writing systems
* **forceException** Force throwing syntax exception if any

Locale validator options

* **filePathPattern** Locale files path (supports node glob pattern)
* **multiFile** Each locale is in separate file in the same folder
* **fileStructure** Structure of locale file content (`polymer` or `angular.flat` file structure)

## Usage

### Locale validator

Validate locale file(s) (e.g. `gulp`)

```javascript
const gulp = require("gulp");
const localizer = require("app-localizer");

gulp.task("validateLocales", function validateLocales(callback) {
  localizer.validateLocales(
    "locales/app1/",
    { multiFile: true, fileStructure: "polymer" },
    undefined,
    result => {
      // result contains missing labels if any
      callback(result);
    }
  );
});

gulp.task("validateMultipleLocales", function validateMultipleLocales(
  callback
) {
  localizer.validateMultipleLocales(
    ["locales/app1/", "locales/app2/"],
    { multiFile: true, fileStructure: "polymer" },
    undefined,
    result => {
      // result contains missing labels if any
      callback(result);
    }
  );
});
```

### Pseudo locale generator

Generate pseudo locale json file from source (`gulp`)

```javascript
const gulp = require("gulp");
const localizer = require("app-localizer");
const rename = require("gulp-rename");

gulp.task("locales", function generatePseudoLocale() {
  gulp
    .src("app/locales/en-us.json")
    .pipe(rename("pseudo.json"))
    .pipe(
      localizer.pseudoLocalize({
        expander: 0.2,
        accents: true,
        rightToLeft: false,
        exclamations: true,
        brackets: true,
        wordexpander: 0.5,
        forceException: false,
        pseudoLocaleName: "en-us"
      })
    )
    .pipe(gulp.dest("dist/locales/"));
});
```

Generate pseudo locale json file from source (`grunt`)

```javascript
const localizer = require("app-localizer");

module.exports = function gruntEntry(grunt) {
  grunt.initConfig({
    pkg: grunt.file.readJSON("package.json"),
    pseudo: {
      options: {
        expander: 0.2,
        exclamations: true,
        brackets: true,
        accents: true,
        rightToLeft: false,
        wordexpander: 0.5,
        forceException: false,
        pseudoLocaleName: "en-us"
      },
      dist: {
        files: {
          "app/locales/pseudo.json": ["app/locales/en-us.json"]
        }
      }
    }
  });

  grunt.registerMultiTask("pseudo", "Pseudolocalize locale", function() {
    const options = this.options();
    this.files.forEach(file => {
      if (file.dest) {
        file.src.forEach(source => {
          const text = grunt.file.read(source);
          const result = localizer.pseudoLocalizeContent(options, text);

          grunt.file.write(file.dest, result);
        });
      }
    });
  });
};
```

Generate pseudo pseudo text (`browser`)

```shell
npm install intl-messageformat-parser
npm install app-localizer
```

```html
	<script type="text/javascript" src="/node_modules/intl-messageformat-parser/dist/parser.js" defer></script>
	<script type="text/javascript" src="/node_modules/app-localizer/localizer.js" defer></script>
	<script type="text/javascript">
		function transform() {
			'use strict';
			const text = AppLocalizer.toPseudoText('some text', { expander: 0.5, accents: true, wordexpander: 0.2 }, IntlMessageFormatParser);
			console.log(text);
		}
	</script>
```

Generate pseudo locale text ([try it](https://runkit.com/58fc19cf15bef7001293bfb4/58fc19cf15bef7001293bfb5))

> transforms
>
> `!"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_abcdefghijklmnopqrstuvwxyz|~`
>
> into
>
> `¡″♯€‰⅋´{}⁎⁺،‐·⁄⓪①②③④⑤⑥⑦⑧⑨∶⁏≤≂≥¿՞ÅƁÇÐÉƑĜĤÎĴĶĻṀÑÖÞǪŔŠŢÛṼŴẊÝŽ⁅∖⁆˄‿‵åƀçðéƒĝĥîĵķļɱñöþǫŕšţûṽŵẋýž¦˞`

```javascript
const localizer = require("app-localizer");

const pseudoText = localizer.toPseudoText(text, {
  expander: 0.2,
  exclamations: true,
  brackets: true,
  accents: true,
  rightToLeft: false,
  wordexpander: 0.5,
  forceException: false
});
```

## Single-File locale example

### file with locales `/locales/locale.json`

```json
{
  "en-us": {
    "label3": "blah3 {token}",
    "label1": "blah1",
    "label2": "blah2",
    "label4": "blah4"
  },
  "de-de": {
    "label3": "blah3 {token}",
    "label1": "blah1",
    "label2": "blah2",
    "label4": "blah4"
  }
}
```

## Multi-File locale example

> locale files should be in the same folder

### file with en-us locale `/locales/en-us.json` (polymer file structure)

```json
{
  "en-us": {
    "label3": "blah3 {token}",
    "label1": "blah1",
    "label2": "blah2",
    "label4": "blah4"
  }
}
```

### file with fr locale `/locales/fr.locale.json` (angular flat file structure)

```json
{
  "label1": "blah1 {{token}}",
  "label2": "blah2",
  "label5": "blah3",
  "label3": "blah4"
}
```

## Change Log

[Change Log](CHANGELOG.md)

## License

[MIT](LICENSE.md)


CHANGELOG.md
# Change Log

## 1.2.2 (March 9, 2018)

* Issue: angular.flat vscode base local handling issue fix

## 1.2.0 (Aug 3, 2017)

* Improvement: Support pseudo generating off json content for web site

## 1.1.19 (July 14, 2017)

* Improvement: 100% code coverage

## 1.1.14 (June 5, 2017)

* Improvement: code coverage
* Improvement: documentation.

## 1.1.13 (May 10, 2017)

* Improvement: added pseudoLocaleName option
* Issue: generate unnecessary `null` in `{pctBlack, number, null}` for pseudo text.

## 1.1.12 (May 9, 2017)

* Improvement: browser support

## 1.1.11 (May 6, 2017)

* Improvement: leverages intl-messageformat-parser for improving pseudo generator

## 1.1.10 (May 4, 2017)

* Bug: fixed several token handling issues for pseudo generator


test/main.js
const localizer = require("../");
const assert = require("assert");
const File = require("vinyl");
const IntlMessageFormat = require("intl-messageformat");

describe("localizer", function() {
	describe("toPseudoText method with", function() {
		it("mapping", function(done) {
			const input =
        " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz|~";
			const output =
        " ¡″#€‰⅋´❨❩⁎⁺،‐·⁄⓪①②③④⑤⑥⑦⑧⑨∶⁏≤≂≥¿՞ÅƁÇÐÉƑĜĤÎĴĶĻṀÑÖÞǪŔŠŢÛṼŴẊÝŽ⁅∖⁆˄‿‵åƀçðéƒĝĥîĵķļɱñöþǫŕšţûṽŵẋýž¦˞";
			assert.equal(localizer.toPseudoText(input, { accents: true }), output);
			done();
		});

		it("no options", function(done) {
			assert.equal(localizer.toPseudoText("some text"), "some text");
			done();
		});

		it("expander", function(done) {
			assert.equal(
				localizer.toPseudoText("some text", { expander: 0.2 }),
				"some text"
			);
			done();
		});

		it("expander (long)", function(done) {
			assert.equal(
				localizer.toPseudoText("some text", { expander: 0.6 }),
				"some some text"
			);
			done();
		});

		it("expander (longest)", function(done) {
			assert.equal(
				localizer.toPseudoText("some text", { expander: 3 }),
				"some some some some text text text text"
			);
			done();
		});

		it("wordexpander", function(done) {
			assert.equal(localizer.toPseudoText("me", { wordexpander: 0.2 }), "me");
			done();
		});

		it("wordexpander (long)", function(done) {
			assert.equal(
				localizer.toPseudoText("some text", { wordexpander: 0.7 }),
				"ssoomme tteexxt"
			);
			done();
		});

		it("wordexpander (longest)", function(done) {
			assert.equal(
				localizer.toPseudoText("some text sometext", { wordexpander: 3 }),
				"sssooooommmmmeee ttteeeeexxxxxttt sssooooommmmeeeeetttteeeexxxxxtt"
			);
			done();
		});

		it("accents", function(done) {
			assert.equal(
				localizer.toPseudoText("some text", { accents: true }),
				"šöɱé ţéẋţ"
			);
			done();
		});

		it("exclamations", function(done) {
			assert.equal(
				localizer.toPseudoText("some text", { exclamations: true }),
				"!!! some text !!!"
			);
			done();
		});

		it("brackets", function(done) {
			assert.equal(
				localizer.toPseudoText("some text", { brackets: true }),
				"[ some text ]"
			);
			done();
		});

		it("rightToLeft token", function(done) {
			assert.equal(
				localizer.toPseudoText("some {token} text", { rightToLeft: true }),
				"\u200F\u202e" + "some {token} text" + "\u202c\u200F"
			);
			done();
		});

		it("missing close token", function(done) {
			assert.equal(
				localizer.toPseudoText("some text {token and {{token}}", {
					expander: 0.5,
					accents: true,
					wordexpander: 0.2
				}),
				"ššöɱé ššöɱé ţţéẋţ ţţéẋţ {{ţöķéñ ååñð ååñð {{{ţöķķéñ}}"
			);
			done();
		});

		it("missing open token", function(done) {
			assert.equal(
				localizer.toPseudoText("some text token} and {{token}} end", {
					expander: 0.5,
					accents: true,
					wordexpander: 0.2
				}),
				"ššöɱé ššöɱé ţţéẋţ ţţöķéñ} ţţöķéñ} ååñð ååñð {{{ţöķķéñ}} ééñð"
			);
			done();
		});

		it("missing token exception", function(done) {
			try {
				localizer.toPseudoText("some text token} and {{token}} end", {
					expander: 0.5,
					accents: true,
					wordexpander: 0.2,
					forceException: true
				});
			} catch (err) {
				assert.equal(err.name, "SyntaxError");
				done();
			}
		});

		it("tokens Formatted Argument", function(done) {
			const input =
        "On {takenDate, date, YYYY} {name} took {pctBlack, number, percent} for {takenDate, time, full} and {pctBlack, number}";
			const output =
        "Öñ Öñ {takenDate, date, YYYY} {name} ţţööķ ţţööķ {pctBlack, number, percent} ƒƒöŕ ƒƒöŕ {takenDate, time, full} ååñð ååñð {pctBlack, number}";
			assert.equal(
				localizer.toPseudoText(input, {
					expander: 0.5,
					accents: true,
					wordexpander: 0.2
				}),
				output
			);
			var msg = new IntlMessageFormat(output);
			assert.equal(
				msg
					.format({
						takenDate: new Date("5/6/2017"),
						name: "petr",
						pctBlack: 0.33
					})
					.replace("GMT", "EDT"),
				"Öñ Öñ 5/6/2017 petr ţţööķ ţţööķ 33% ƒƒöŕ ƒƒöŕ 12:00:00 AM EDT ååñð ååñð 0.33"
			);
			done();
		});

		it("tokens {plural} Format", function(done) {
			const input =
        "You have {itemCount, plural, =0 {no items} one {1 item} other {{itemCount} items}}.";
			const output =
        "ÝÝöû ÝÝöû ĥĥåṽé {itemCount, plural, =0 {ñö ñö îîţéɱš} one {① ① îîţéɱ} other {{itemCount} îîţéɱš îîţéɱš}}· ·";
			assert.equal(
				localizer.toPseudoText(input, {
					expander: 0.5,
					accents: true,
					wordexpander: 0.2
				}),
				output
			);
			var msg = new IntlMessageFormat(output);
			assert.equal(
				msg.format({ itemCount: 0 }),
				"ÝÝöû ÝÝöû ĥĥåṽé ñö ñö îîţéɱš· ·"
			);
			done();
		});

		it("tokens {select} Format", function(done) {
			const input =
        "{gender, select, male {He {taxRate, number, percent}} female {She} other {They}} will respond shortly.";
			const output =
        "{gender, select, male {Ĥé Ĥé {taxRate, number, percent}} female {ŠŠĥé ŠŠĥé} other {ŢŢĥéý ŢŢĥéý}} ŵŵîļļ ŵŵîļļ ŕŕéšþöñð ŕŕéšþöñð ššĥöŕţţļý·";
			assert.equal(
				localizer.toPseudoText(input, {
					expander: 0.5,
					accents: true,
					wordexpander: 0.2
				}),
				output
			);
			var msg = new IntlMessageFormat(output);
			assert.equal(
				msg.format({ gender: "female" }),
				"ŠŠĥé ŠŠĥé ŵŵîļļ ŵŵîļļ ŕŕéšþöñð ŕŕéšþöñð ššĥöŕţţļý·"
			);
			done();
		});

		it("tokens {selectordinal} Format", function(done) {
			const input =
        "It's my cat's {year, selectordinal, offset:1 one {#st} two {#nd} few {#rd} other {#th}} birthday!";
			const output =
        "ÎÎţ´š ÎÎţ´š ɱý ɱý ççåţ´š {year, selectordinal, offset:1 one {#šţ šţ} two {#ñð ñð} few {#ŕð ŕð} other {#ţĥ ţĥ}} ƀƀîŕţĥĥðåý¡ ƀƀîŕţĥĥðåý¡";
			assert.equal(
				localizer.toPseudoText(input, {
					expander: 0.5,
					accents: true,
					wordexpander: 0.2
				}),
				output
			);
			var msg = new IntlMessageFormat(output);
			assert.equal(
				msg.format({ year: 2 }),
				"ÎÎţ´š ÎÎţ´š ɱý ɱý ççåţ´š 1šţ šţ ƀƀîŕţĥĥðåý¡ ƀƀîŕţĥĥðåý¡"
			);
			done();
		});
	});

	describe("pseudoLocalize method with", function() {
		it("pseudo localize polymer", function(done) {
			const result =
        '{\n	"fr-be": {\n		"label1": "sssooommmmee {token1} ttteeexxxxtt {token2}"\n	}\n}';
			assert.equal(
				localizer.pseudoLocalizeContent(
					{ wordexpander: 2, pseudoLocaleName: "fr-be" },
					'{ "en-us": { "label1": "some {token1} text {token2}" } }'
				),
				result
			);
			done();
		});

		it("pseudo localize angular.flat", function(done) {
			const result =
        '{\n	"label1": "sssooommmmee {token1} ttteeexxxxtt {token2}"\n}';
			assert.equal(
				localizer.pseudoLocalizeContent(
					{ wordexpander: 2, format: "angular.flat" },
					'{ "label1": "some {token1} text {token2}" }'
				),
				result
			);
			done();
		});

		it("pseudo localize gulp", function(done) {
			const result =
        '{\n	"pseudo": {\n		"label1": "sssooommmmee {token1} ttteeexxxxtt {token2}"\n	}\n}';
			const localeFile = new File({
				contents: new Buffer(
					'{ "en-us": { "label1": "some {token1} text {token2}" } }'
				)
			});
			const localizerPlugin = localizer.pseudoLocalize({ wordexpander: 2 });
			localizerPlugin.write(localeFile);
			localizerPlugin.once("data", function(file) {
				assert(file.isBuffer());
				assert.equal(file.contents, result);
				done();
			});
		});

		it("empty pseudo localize gulp", function(done) {
			const localeFile = new File();
			const localizerPlugin = localizer.pseudoLocalize({ wordexpander: 2 });
			localizerPlugin.write(localeFile);
			localizerPlugin.once("data", function(file) {
				assert.ifError(file.isBuffer());
				done();
			});
		});
	});

	describe("validateLocales method with", function() {
		it("validate path", function(done) {
			const result = {
				"en-us": { label1: ["de-de", "be-by"], label3: ["de-de", "be-by"] },
				"be-by": { label2: ["en-us"] },
				"de-de": { label2: ["en-us"] }
			};
			const input =
        '{ "en-us": { "label1": "some {token1} text {{token2}}", "label3": "some {token1} text {{token2}}" }, "de-de": { "label2": "some {token1} text {{token2}}" }, "be-by": { "label2": "some {token1} text {{token2}}" } }';
			assert.deepEqual(
				localizer.validateLocalesContent(JSON.parse(input)),
				result
			);
			done();
		});

		it("single file", function(done) {
			const expectedResult = {
				"en-us": { label1: ["de-de"] },
				"de-de": { label2: ["en-us"] }
			};
			localizer.validateLocales(
				"locales/test1/en-us.json",
				{ multiFile: false },
				undefined,
				result => {
					assert.deepEqual(result, expectedResult);
					done();
				}
			);
		});

		it("missing file", function() {
			assert.doesNotThrow(() => {
				localizer.validateLocales(
					"locales/test1/en-us_missing.json",
					{ multiFile: false },
					undefined,
					() => {}
				);
				localizer.validateLocales(
					"locales/test_missing/",
					{ multiFile: true },
					undefined,
					() => {}
				);
			});
		});

		it("polymer multi-file", function(done) {
			const expectedResult = {
				"en-us": { label1: ["de-de"] },
				"de-de": { label2: ["en-us"] }
			};
			localizer.validateLocales(
				"locales/test2/",
				{ multiFile: true },
				undefined,
				result => {
					assert.deepEqual(result, expectedResult);
					done();
				}
			);
		});

		it("base (polymer multi-file)", function(done) {
			const expectedResult = { "en-us": { label1: ["de-de"] } };
			localizer.validateLocales(
				"locales/test2/",
				{ multiFile: true },
				{ "en-us": { label1: "some text" } },
				result => {
					assert.deepEqual(result, expectedResult);
					done();
				}
			);
		});

		it("angular.flat multi-file", function(done) {
			const expectedResult = {
				"en-us": { label1: ["de-de"] },
				"de-de": { label2: ["en-us"] }
			};
			localizer.validateLocales(
				"locales/test3/",
				{ multiFile: true, fileStructure: "angular.flat" },
				undefined,
				result => {
					assert.deepEqual(result, expectedResult);
					done();
				}
			);
		});

		it("angular.flat multi-file with base", function(done) {
			const expectedResult = {
				base: { label2: ["en-us"] }
			};
			localizer.validateLocales(
				"locales/test3/",
				{ multiFile: true, fileStructure: "angular.flat" },
				{
					label2: "some text"
				},
				result => {
					assert.deepEqual(result, expectedResult);
					done();
				}
			);
		});

		it("multi path (angular.flat multi-file)", function(done) {
			const expectedResult = {
				"locales/test3/": {
					"en-us": { label1: ["de-de"] },
					"de-de": { label2: ["en-us"] }
				},
				"locales/test4/": {
					"en-us": { label3: ["de-de"] },
					"de-de": { label2: ["en-us"] }
				}
			};
			localizer.validateMultipleLocales(
				["locales/test3/", "locales/test4/"],
				{ multiFile: true, fileStructure: "angular.flat" },
				undefined,
				result => {
					assert.deepEqual(result, expectedResult);
					done();
				}
			);
		});

		it("promise multi path (angular.flat multi-file)", function(done) {
			const expectedResult = {
				"locales/test3/": {
					"en-us": { label1: ["de-de"] },
					"de-de": { label2: ["en-us"] }
				},
				"locales/test4/": {
					"en-us": { label3: ["de-de"] },
					"de-de": { label2: ["en-us"] }
				}
			};
			const promise = localizer.validateMultipleLocales(
				["locales/test3/", "locales/test4/"],
				{ multiFile: true, fileStructure: "angular.flat" }
			);
			promise.catch(result => {
				assert.deepEqual(result, expectedResult);
				done();
			});
		});

		it("promise multi path valid (angular.flat multi-file)", function(done) {
			const promise = localizer.validateMultipleLocales(
				["locales/test5/", "locales/test5/"],
				{ multiFile: true, fileStructure: "angular.flat" }
			);
			promise.then(() => {
				done();
			});
		});
	});
});


localizer.js
(function() {
	'use strict';

	const appLocalizer = (function() {
		'use strict';
		const mapping = {
			" ": "\u2003",
			"!": "\u00a1",
			"\"": "\u2033",
			"$": "\u20ac",
			"%": "\u2030",
			"&": "\u214b",
			"'": "\u00b4",
			"(": "\u2768",
			")": "\u2769",
			"*": "\u204e",
			"+": "\u207a",
			",": "\u060c",
			"-": "\u2010",
			".": "\u00b7",
			"/": "\u2044",
			"0": "\u24ea",
			"1": "\u2460",
			"2": "\u2461",
			"3": "\u2462",
			"4": "\u2463",
			"5": "\u2464",
			"6": "\u2465",
			"7": "\u2466",
			"8": "\u2467",
			"9": "\u2468",
			":": "\u2236",
			";": "\u204f",
			"<": "\u2264",
			"=": "\u2242",
			">": "\u2265",
			"?": "\u00bf",
			"@": "\u055e",
			"A": "\u00c5",
			"B": "\u0181",
			"C": "\u00c7",
			"D": "\u00d0",
			"E": "\u00c9",
			"F": "\u0191",
			"G": "\u011c",
			"H": "\u0124",
			"I": "\u00ce",
			"J": "\u0134",
			"K": "\u0136",
			"L": "\u013b",
			"M": "\u1e40",
			"N": "\u00d1",
			"O": "\u00d6",
			"P": "\u00de",
			"Q": "\u01ea",
			"R": "\u0154",
			"S": "\u0160",
			"T": "\u0162",
			"U": "\u00db",
			"V": "\u1e7c",
			"W": "\u0174",
			"X": "\u1e8a",
			"Y": "\u00dd",
			"Z": "\u017d",
			"[": "\u2045",
			"\\": "\u2216",
			"]": "\u2046",
			"^": "\u02c4",
			"_": "\u203f",
			"`": "\u2035",
			"a": "\u00e5",
			"b": "\u0180",
			"c": "\u00e7",
			"d": "\u00f0",
			"e": "\u00e9",
			"f": "\u0192",
			"g": "\u011d",
			"h": "\u0125",
			"i": "\u00ee",
			"j": "\u0135",
			"k": "\u0137",
			"l": "\u013c",
			"m": "\u0271",
			"n": "\u00f1",
			"o": "\u00f6",
			"p": "\u00fe",
			"q": "\u01eb",
			"r": "\u0155",
			"s": "\u0161",
			"t": "\u0163",
			"u": "\u00fb",
			"v": "\u1e7d",
			"w": "\u0175",
			"x": "\u1e8b",
			"y": "\u00fd",
			"z": "\u017e",
			"|": "\u00a6",
			"~": "\u02de"
		}
		/**
		 * @typedef GeneratorOptions
		 * @type {Object}
		 * @property {number} expander Sentance expanding factor (0.3 = 30%).
		 * @property {number} wordexpander Word expanding factor (0.5 = 50%).
		 * @property {boolean} accents Convert letter to its accent version.
		 * @property {boolean} exclamations Enclose in exclamations.
		 * @property {boolean} brackets Enclose in brackets.
		 * @property {boolean} rightToLeft Left-to-Right.
		 * @property {boolean} forceException throw syntax exception if any.
		 *
		 * @typedef PseudoLocalizerOptions
		 * @property {number} expander Sentance expanding factor (0.3 = 30%).
		 * @property {number} wordexpander Word expanding factor (0.5 = 50%).
		 * @property {boolean} accents Convert letter to its accent version.
		 * @property {boolean} exclamations Enclose in exclamations.
		 * @property {boolean} brackets Enclose in brackets.
		 * @property {boolean} rightToLeft Left-to-Right.
		 * @property {string} format Structure of locale file content (polymer, angular.flat).
		 *
		 * @typedef ValidateOptions
		 * @property {boolean} multiFile Locale is localed in separate file.
		 * @property {string} fileStructure Structure of locale file content (polymer, angular.flat).
		*/

		/**
		 * Transform sentences into expanded sentences with accents.
		 *
		 * @param {string} text Sentences.
		 * @param {GeneratorOptions} options Generator options.
		 * @returns {string} Pseudo generated sentences.
		 */
		function transformSentences(text, options) {
			if(options && text && text !== ' ') {
				let words = text.split(' ');
				const expand = function expand(items, factor, callback) {
					const extraCount = Math.round(items.length * factor);
					let extraPosition = 0;
					for (let i = 0; i < extraCount; i += 1) {
						const position = Math.round(extraPosition);
						const expandedPosition = Math.round((items.length + extraCount - 1) / (items.length - 1) * extraPosition) + 1;
						callback(expandedPosition, items[position]);
						extraPosition += (items.length - 1) / extraCount;
					}
				};

				if(options.expander) {
					const extendedWords = words.slice(0);
					expand(words, options.expander, (position, item) => {
						extendedWords.splice(position, 0, item);
					});
					words = extendedWords;
				}

				if(options.wordexpander) {
					const expandedWords = [];
					words.forEach((word) => {
						let expandedWord = word;
						expand(word, options.wordexpander, (position, item) => {
							expandedWord = `${expandedWord.substring(0, position)}${item}${expandedWord.substring(position)}`;
						});
						expandedWords.push(expandedWord);
					});
					words = expandedWords;
				}

				if(options.accents) {
					const accentedWords = [];
					words.forEach((word) => {
						word = [...word].map(char => mapping[char] ? mapping[char] : char).join('');
						accentedWords.push(word);
					});
					words = accentedWords;
				}

				text = words.join(' ');
			}
			return text;
		};

		/**
		 * Walk through parsed AST.
		 *
		 * @param {Object} node parsed AST.
		 * @param {Object[]} parts List of text parts.
		 */
		function walkAST(node, parts) {
			switch(node.type) {
				case 'messageTextElement':
					// Hashtag is a key word
					const hash = node.value.split('#');
					if(hash.length > 1) {
						hash.forEach((part) => {
							parts.push({ token: false, text: part });
							parts.push({ token: true, text: '#' });
						});
						parts.pop();
					} else {
						parts.push({ token: false, text: node.value });
					}
					break;
				case 'messageFormatPattern':
					node.elements.forEach((subnode) => walkAST(subnode, parts));
					break;
				case 'argumentElement':
					parts.push({ token: true, text: `{${node.id}` });
					if(node.format) {
						walkAST(node.format, parts);
					}
					parts.push({ token: true, text: '}' });
					break;
				case 'pluralFormat':
					parts.push({ token: true, text: `, ${node.ordinal ? 'selectordinal' : 'plural'},${node.offset ? ` offset:${node.offset}`: ''}` });
					node.options.forEach((subnode) => walkAST(subnode, parts));
					break;
				case 'selectFormat':
					parts.push({ token: true, text: `, select,` });
					node.options.forEach((subnode) => walkAST(subnode, parts));
					break;
				case 'optionalFormatPattern':
					parts.push({ token: true, text: ` ${node.selector} {` });
					walkAST(node.value, parts);
					parts.push({ token: true, text: '}' });
					break;
				case 'dateFormat':
				case 'numberFormat':
				case 'timeFormat':
					parts.push({ token: true, text: `, ${node.type.substring(0, node.type.length - 'Format'.length)}${node.style ? `, ${node.style}` : ''}` });
					break;
			}
		};

		/**
		 * Generates pseudo text.
		 *
		 * @param {string} text Input text.
		 * @param {GeneratorOptions} options Generator options
		 * @returns {string} Pseudo generated text
		 */
		function toPseudoText(text, options, messageParser) {
			let result = text;
			if(options) {
				let message = undefined;
				try {
					message = messageParser.parse(text);
				} catch(err) {
					if(options.forceException) {
						throw err;
					}
				}

				const parts = [];
				if(message) {
					walkAST(message, parts, options);
				} else {
					parts.push({ text });
				}

				for (let index = 0; index < parts.length; index++) {
					if(!parts[index].token && parts[index].text !== ' ') {
						// Text part can start or end with space
						const startsFromSpace = parts[index].text[0] === ' ';
						const endsWithSpace = parts[index].text[parts[index].text.length - 1] === ' ';

						parts[index].text = `${startsFromSpace ? ' ': ''}${transformSentences(parts[index].text.trim(), options)}${endsWithSpace ? ' ': ''}`;
					}
				}

				if(options.exclamations) {
					parts.splice(0, 0, { text: '!!! ' });
					parts.push({ text: ' !!!' });
				}

				if(options.brackets) {
					parts.splice(0, 0, { text: '[ ' });
					parts.push({ text: ' ]' });
				}

				result = parts.map((part) => part.text).join('');

				if(options.rightToLeft) {
					const RLO = '\u202e';
					const PDF = '\u202c';
					const RLM = '\u200F';
					result = RLM + RLO + result + PDF + RLM;
				}
			}
			return result;
		};

		/**
		 * Generates pseudo locale.
		 *
		 * @param {PseudoLocalizerOptions} options Generator options.
		 * @param {string} text Input json file content.
		 * @returns {string} Pseudo generated json content.
		 */
		function pseudoLocalizeContent(options, text, messageParser) {
			let locale = JSON.parse(text);
			const localename = options.format === 'angular.flat' ? '' : Object.keys(locale)[ 0 ];
			const transformed = {};

			if (localename) {
				locale = locale[localename];
			}

			Object.keys(locale).forEach((key) => {
				transformed[key] = toPseudoText(locale[key], options, messageParser);
			} );

			let result = transformed;
			if (localename) {
				result = {};
				result[options.pseudoLocaleName ? options.pseudoLocaleName : 'pseudo'] = transformed;
			}

			return JSON.stringify( result, null, '\t' );
		};
		return {
			toPseudoText: toPseudoText,
			pseudoLocalizeContent: pseudoLocalizeContent
		};
	})();

	this['AppLocalizer'] = appLocalizer;
}).call(this);